### PR TITLE
Nearest station

### PR DESCRIPTION
**Pull request by liliannaszabo**

### CODE REVIEW

File: src/components/Dropdown.tsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import Button from './Button';
+
+type DropdownProps = {
+    onChange: (option: React.SetStateAction<Station | undefined> ) => void;
+    stationsList: Station[];
+    selectedOption: Station | undefined;
+}
+
+const DropdownMenu: React.FC<DropdownProps>  = ({ onChange, selectedOption, stationsList }) => {
+    const handleSelectChange = (option?: Station | null | undefined) => {
```
Comment (WendiF) [inline src/components/Dropdown.tsx]: Do we need `?` and `null` and `undefined`?

File: src/components/StationPrompt.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import JourneysList from './JourneysList';
+
+const StationPrompt: () => JSX.Element = () => {
+    const [stationFrom, setStationFrom] = useState<Station>();
+    const [stationTo, setStationTo] = useState<Station>();
+    const [journeyData, setJourneyData] = useState<Journey[]>([]);
+    const [stationsList, setStationsList] = useState<Station[]>([]);
+
+    const getTrains = () => {
+        if(stationFrom && stationTo){
```
Comment (WendiF) [inline src/components/StationPrompt.tsx]: spacing 
```suggestion
        if (stationFrom && stationTo) {
```

File: src/helpers/ApiCallHelper.tsx [hunk 3]
```diff
@@ -1,7 +1,28 @@
...
+    });
+
+    return fetch(`https://mobile-api-softwire2.lner.co.uk/v1/fares?${queryParams}`, {
+        headers: {
+            'X-API-KEY': `${process.env.REACT_APP_X_API_KEY}`,
+        },
+    });
+    // } else {
+    //    
+    // }
```
Comment (WendiF) [inline src/helpers/ApiCallHelper.tsx]: ?

File: src/helpers/LocationHelper.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import { Location } from '../models/Location';
+import { Station } from '../models/Station';
+
+export const getLocation = (): Promise<Location> => {
+    return new Promise((resolve, reject) => {
+        if( navigator.geolocation){
```
Comment (WendiF) [inline src/helpers/LocationHelper.tsx]:
```suggestion
        if (navigator.geolocation) {
```

File: src/helpers/LocationHelper.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+export const getDistance = (x: number, y: number, x2: number, y2: number) => {
+    return Math.sqrt((y-y2)**2 + (x-x2)**2 );
+};
+
+export const getNearestStation = (location: Location, stations: Station[]) => {
+    let closestDistance = Infinity;
+    let closestStation = undefined;
+    for(const station of stations) {
+        const distance = getDistance(location.latitude,location.longitude,station.latitude,station.longitude);
+        if(distance<closestDistance){
```
Comment (WendiF) [inline src/helpers/LocationHelper.tsx]:
```suggestion
    for (const station of stations) {
        const distance = getDistance(location.latitude, location.longitude, station.latitude, station.longitude);
        if (distance < closestDistance) {
```

File: src/mappers/ResponseMapper.tsx [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        station.label = station.name;
+    });
+    return stations;
+};
+
+//data => data.json().then(d => console.log(d.outboundJourneys)))
+//
+// .then(response => response.json()
+//     .then(jsonData => jsonData.map((station: StationsListStation) =>
+//         stations.push({ label: station.name, value: station.crs }))));
```
Comment (WendiF) [inline src/mappers/ResponseMapper.tsx]: ?

File: src/models/Station.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    nlc: string;
+    latitude: number;
+    longitude: number;
+    value: string;
+    label: string;
+}
+
+// export interface DropdownStationOption {
+//
+// }
```
Comment (WendiF) [inline src/models/Station.tsx]: Make sure you have a look at your changes before committing them

File: src/models/Journey.tsx [hunk 8]
```diff
@@ -0,0 +1,13 @@
+
+export type Journey = {
+    arrivalTime: string;
+    departureTime: string;
+    destinationStation: JourneyStation;
```
Comment (WendiF) [inline src/models/Journey.tsx:5]: Why does this not have an originStation?

File: src/components/StationPrompt.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        });
+        
+    }
+
+    ,[]);
+
+    return (
+        <div>
+            <DropdownMenu onChange = { setStationFrom } stationsList = { stationsList } selectedOption = { stationFrom } />
+            <DropdownMenu onChange = { setStationTo } stationsList = { stationsList } selectedOption = { stationTo }  />
```
Comment (WendiF) [inline src/components/StationPrompt.tsx]: You use a different naming convention for this elsewhere - can we keep it consistent i.e. change these to `setOriginStation` and `setDestinationStation`?

File: src/components/JourneysList.tsx [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    }, [journeyData]);
+
+    return (
+        <div>
+            <tbody>
+                <tr>
+                    <th>Depart Time</th>
+                    <th>Arrival Time</th>
+                </tr>
+                {journeyData.map(( item, index ) => (
```
Comment (WendiF) [inline src/components/JourneysList.tsx]: Specify what the item is here with something more descriptive please

File: src/components/Dropdown.tsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
+import React, { useEffect, useState } from 'react';
+import { useParams } from 'react-router-dom';
+import Select  from 'react-select';
+import { stat } from 'fs';
+import { fetchStations } from '../helpers/ApiCallHelper';
+import { responseMapper } from '../mappers/ResponseMapper';
+import {  Station } from '../models/Station';
```
Comment (WendiF) [inline src/components/Dropdown.tsx]:
```suggestion
import { Station } from '../models/Station';
```

File: src/components/Dropdown.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        <div>
+            <label htmlFor = "dropdown">Select an option:</label>
+            <Select options = { stationsList }
+                onChange = { (option: Station | null | undefined) =>  handleSelectChange(option) }
+                value = { selectedOption }/>
+        </div>
+    );
+};
+
+export default DropdownMenu;
```
Comment (WendiF) [inline src/components/Dropdown.tsx]: `DropdownMenu` is generic but it has a prop called `stationsList` which is specific - I have a weak preference for keeping things specific when you can be (i.e. renaming to `StationSelectDropdown` or something), over having a generic component when you don't need it, but you should keep to one or the other. If you have station selection specific logic in this class, you could also name it accordingly. If you want to keep it generic, don't use logic/naming in it that can only be used for a specific scenario.

Also, the class and the file name do not match

File: src/App.tsx [hunk 13]
```diff
@@ -15,6 +16,7 @@ const App = () => (
                 </Route>
             </Routes>
             <footer>
+                <StationPrompt />
```
Comment (WendiF) [inline src/App.tsx]: This probably shouldn't be in the footer

**Review by WendiF — CHANGES_REQUESTED** [review]

File: src/components/Dropdown.tsx (hunk 1 above)
Comment (liliannaszabo) [reply src/components/Dropdown.tsx]: When remuving it a " Argument of type 'Station | undefined' is not assignable to parameter of type 'Station'" error comes up on line 24 in the useEffect

File: src/components/StationPrompt.tsx (hunk 2 above)
Comment (liliannaszabo) [reply src/components/StationPrompt.tsx]: changed now

File: src/helpers/ApiCallHelper.tsx (hunk 3 above)
Comment (liliannaszabo) [reply src/helpers/ApiCallHelper.tsx]: got rid of it now

File: src/helpers/LocationHelper.tsx (hunk 4 above)
Comment (liliannaszabo) [reply src/helpers/LocationHelper.tsx]: Changed it now

File: src/helpers/LocationHelper.tsx (hunk 5 above)
Comment (liliannaszabo) [reply src/helpers/LocationHelper.tsx]: changed now

File: src/models/Station.tsx (hunk 7 above)
Comment (liliannaszabo) [reply src/models/Station.tsx]: Changed it now

File: src/models/Journey.tsx (hunk 8 above)
Comment (liliannaszabo) [reply src/models/Journey.tsx:5]: added an originStation

File: src/components/StationPrompt.tsx (hunk 9 above)
Comment (liliannaszabo) [reply src/components/StationPrompt.tsx]: Changed it everywhere

File: src/components/JourneysList.tsx (hunk 10 above)
Comment (liliannaszabo) [reply src/components/JourneysList.tsx]: changed it to journey

File: src/components/Dropdown.tsx (hunk 12 above)
Comment (liliannaszabo) [reply src/components/Dropdown.tsx]: Changed both the name and the filename

File: src/mappers/ResponseMapper.tsx (hunk 6 above)
Comment (liliannaszabo) [reply src/mappers/ResponseMapper.tsx]: Got rid of it

File: src/components/Dropdown.tsx (hunk 11 above)
Comment (liliannaszabo) [reply src/components/Dropdown.tsx]: Changed it now

File: src/App.tsx (hunk 13 above)
Comment (liliannaszabo) [reply src/App.tsx]: Changed it

File: src/components/Dropdown.tsx (hunk 1 above)
Comment (liliannaszabo) [reply src/components/Dropdown.tsx]: I am able to get rid of ?

**Review by WendiF — APPROVED** [review]

**Pull request merged**